src/common/store/background-epics.ts

```diff
@@ -16,15 +30,53 @@ interface FluxStandardAction {
 export const locateEpic: Epic<FluxStandardAction> = action$ => action$.pipe(
   ofType('API_READ'),
   filter(action => action.payload.endpoint === endpoints.ANNOTATIONS),
-  mergeMap(async (action) => {
-    const tabId = retrieveActionTab(action);
-    const locationData = await tabLocateAnnotations(tabId, action.payload.data);
-    syncBadgeWithAnnotations(locationData, tabId);
-    const newAction = locateAnnotations(locationData);
-    return syncTabMark(action, newAction);
-  }),
+  mergeMap(action => defer(
+    async () => {
+      const tabId = retrieveActionTab(action);
+      const locationData = await tabLocateAnnotations(tabId, action.payload.data);
+      syncBadgeWithAnnotations(locationData, tabId);
+      const newAction = locateAnnotations(locationData);
+      return syncTabMark(action, newAction);
+    },
+  )),
+);
+
+export const processAuthenticationEpic: Epic<FluxStandardAction, any, IState> = (action$, state$) => (
+  action$.pipe(
+    ofType(SET_AUTH_PROC_STAGE),
+    filter(() => !selectUser(state$.value)),
+    filter(action => action.payload.stage === AuthProcStages.initiated),
+    mergeMap( (action): Observable<FluxStandardAction> =>
+      from(authenticate(action.payload.provider)).pipe(
+        mergeMap( loginData =>
+          loginData ? of(
+            completeAuthProc(),
+            userDataNew(loginData),
+          ) : of(
+            cancelAuthProc('Anulowano autoryzację'),
+          ),
+        ),
```

Comment:
One day I'll get round to understanding this :)

Reply:
I tried to describe it, but gave up after seeing how much of RxJS docs I actually copy ;) We could talk about it for 10mins – it would be the fastest way :)